File: lollys-lotto-ts-sdk/src/codegen/instructions/closeUserMetadata.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CloseUserMetadataAccounts {
  user: PublicKey
  userMetadata: PublicKey
  eventEmitter: PublicKey
  systemProgram: PublicKey
}

export function closeUserMetadata(
  accounts: CloseUserMetadataAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.user, isSigner: true, isWritable: true },
    { pubkey: accounts.userMetadata, isSigner: false, isWritable: true },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([4, 177, 246, 70, 175, 75, 175, 161])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}

File: lollys-lotto-rust-sdk/src/instructions/user/close_user_metadata.rs
use crate::instructions::*;

pub fn close_user_metadata(
    user: &Pubkey,
    user_metadata: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = lollys_lotto::instruction::CloseUserMetadata {}.data();

    let accounts = lollys_lotto::accounts::CloseUserMetadata {
        user: *user,
        user_metadata: *user_metadata,
        event_emitter: *event_emitter,
        system_program: system_program::ID,
    }
    .to_account_metas(None);

    Instruction {
        program_id: lollys_lotto::id(),
        accounts,
        data,
    }
}